### PR TITLE
[post training]  define llama stack post training dataset format

### PR DESCRIPTION
**Pull request by SLR722**
## context
In this PR, we defined 2 llama stack dataset formats (instruct, dialog)

- For instruct dataset format, the column schema will be [chat_completion_input, expected_answer], which is consistent with the eval data format. This dataset format is the abstract of single turn QA style post training data
- For dialog dataset format, the column schema will be [dialog], which is a list of user messages and assistant messages that interleave together. During training, the whole list will be the model input and the loss is calculated on assistant messages only. This dataset format is the abstract of multi turn chat style post training data

## changes
- defined the 2 llama stack dataset formats
- an adapter to convert llama stack dataset format to torchtune dataset format
- move dataset format validation to post training level instead of torchtune level since it's not specific to torchtune
- add localfs as datasetio provider

## test 
instruct format
- use https://huggingface.co/datasets/llamastack/evals as dataset and the training works as expected
Screenshot 2025-01-09 at 5 15 14 PM

- use my generated local dataset and the training works as expected

Screenshot 2025-01-09 at 5 19 11 PM

dialog format
- use my generated local dataset and the training works as expected
Screenshot 2025-01-09 at 5 23 16 PM

### CODE REVIEW

File: llama_stack/providers/inline/post_training/torchtune/common/utils.py [hunk 1]
```diff
@@ -56,6 +70,13 @@ class DatasetSchema(BaseModel):
...
 
+DATA_FORMATS: Dict[str, Transform] = {
+    "alpaca": AlpacaToMessages,
+    "instruct": InputOutputToMessages,
+    "chat_sharegpt": ShareGPTToMessages,
+    "chat_openai": OpenAIToMessages,
+}
+
 
 EXPECTED_DATASET_SCHEMA = DatasetSchema(
```
Comment (yanxi0830) [inline llama_stack/providers/inline/post_training/torchtune/common/utils.py:81]: Might be helpful to use DataSchemaValidatorMixin for data schema validation https://github.com/meta-llama/llama-stack/blob/main/llama_stack/providers/utils/common/data_schema_validator.py#L65

File: llama_stack/providers/inline/post_training/torchtune/common/utils.py [hunk 2]
```diff
@@ -11,26 +11,37 @@
...
+    ShareGPTToMessages,
+)
 
 from torchtune.models.llama3 import llama3_tokenizer, lora_llama3_8b
 from torchtune.models.llama3._tokenizer import Llama3Tokenizer
 from torchtune.models.llama3_2 import lora_llama3_2_3b
+from torchtune.modules.transforms import Transform
 
 
 class ColumnName(Enum):
```
Comment (yanxi0830) [inline llama_stack/providers/inline/post_training/torchtune/common/utils.py:38]: I think this enum class should be merged with https://github.com/meta-llama/llama-stack/blob/485476c29a20be196d1a5e7c4208a13d12a250b6/llama_stack/providers/utils/common/data_schema_validator.py#L19-L25

File: llama_stack/providers/inline/post_training/torchtune/common/utils.py (hunk 1 above)
Comment (yanxi0830) [reply llama_stack/providers/inline/post_training/torchtune/common/utils.py:81]: Synced offline: we can keep logic for validation in common utils, but there's no value in having DataSchemaValidatorMixin. Refactor in https://github.com/meta-llama/llama-stack/pull/720

**Review by qiansun3 — APPROVED** [review]

File: llama_stack/providers/inline/post_training/common/validator.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+EXPECTED_DATASET_SCHEMA = {
+    "instruct": [
+        {
+            ColumnName.chat_completion_input.value: ChatCompletionInputType(),
+            ColumnName.expected_answer.value: StringType(),
+        }
+    ],
+    "dialog": [
+        {
+            ColumnName.dialog.value: StringType(),
```
Comment (ashwinb) [inline llama_stack/providers/inline/post_training/common/validator.py:28]: I have now realized our type system in `apis/common/type_system.py` is not complete at all. We should be able to specify `ListType(StringType())` or `ListType(MessageType())` so we can automatically validate that the json formatted column is correct. Otherwise we are putting all these checks deep inside post-training where they do _not_ belong.

we got around that by introducing a "special" type called `ChatCompletionInputType()` --- why is that not the appropriate value here?

Comment (SLR722) [reply llama_stack/providers/inline/post_training/common/validator.py:28]: My concern with using ChatCompletionInputType() here is that it might be confusing to users due to its semantics. Do you think we could create a new specialized type called DialogType()?

Comment (ashwinb) [reply llama_stack/providers/inline/post_training/common/validator.py:28]: @SLR722 yeah that sounds better to me

Comment (SLR722) [reply llama_stack/providers/inline/post_training/common/validator.py]: addressed

File: llama_stack/apis/common/type_system.py [hunk 4]
```diff
@@ -54,6 +54,11 @@ class AgentTurnInputType(BaseModel):
     type: Literal["agent_turn_input"] = "agent_turn_input"
 
 
+class DialogType(BaseModel):
+    # expects List[Message] for messages
```
Comment (ashwinb) [inline llama_stack/apis/common/type_system.py:58]: I think adding a comment here which says that "this type semantically contains the output label whereas ChatCompletionInputType does not" would be good

**Review by ashwinb — APPROVED** [review]
thanks
